src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.EasyRequest.cs

```diff
@@ -0,0 +1,373 @@
+            {
+                Debug.Assert(error != null, "Expected non-null exception");
+
+                var oce = error as OperationCanceledException;
+                if (oce != null)
+                {
+                    TrySetCanceled(oce.CancellationToken);
+                }
+                else
+                {
+                    TrySetException(error as HttpRequestException ?? CreateHttpRequestException(error));
+                }
+                // There's not much we can reasonably assert here about the result of TrySet*.
+                // It's possible that the task wasn't yet completed (e.g. a failure while initiating the request),
+                // it's possible that the task was already completed as success (e.g. a failure sending back the response),
+                // and it's possible that the task was already completed as failure (e.g. we handled the exception and
+                // faulted the task, but then tried to fault it again while finishing processing in the main loop).
+
+                // Make sure the exception is available on the response stream so that it's propagated
+                // from any attempts to read from the stream.
+                ResponseMessage.ContentStream.SignalFailure(ExceptionDispatchInfo.Capture(error));
+            }
+
+            public void Cleanup() // not called Dispose because the request may still be in use after it's cleaned up
+            {
+                ResponseMessage.ContentStream.SignalComplete(); // No more callbacks so no more data
+                // Don't dispose of the ResponseMessage.ContentStream as it may still be in use
+                // by code reading data stored in the stream.
+
+                // Dispose of the input content stream if there was one.  Nothing should be using it any more.
+                if (RequestContentStream != null)
+                    RequestContentStream.Dispose();
+
+                // Dispose of the underlying easy handle.  We're no longer processing it.
+                if (_easyHandle != null)
+                    _easyHandle.Dispose();
+
+                // Dispose of the request headers if we had any.  We had to keep this handle
+                // alive as long as the easy handle was using it.  We didn't need to do any
+                // ref counting on the safe handle, though, as the only processing happens
+                // in Process, which ensures the handle will be rooted while libcurl is
+                // doing any processing that assumes it's valid.
+                if (_requestHeaders != null)
+                    _requestHeaders.Dispose();
+            }
+
+            private void SetUrl()
+            {
+                SetCurlOption(CURLoption.CURLOPT_URL, _requestMessage.RequestUri.AbsoluteUri);
+            }
+
+            [Conditional("CURLHANDLER_VERBOSE")]
+            private void SetDebugging()
+            {
+                SetCurlOption(CURLoption.CURLOPT_VERBOSE, 1L);
```

Comment:
Potentially.  My understanding is that if you enable verbose but not supply a custom debug function, you essentially get a debug function that writes to stderr the data associated with any CURLINFO_TEXT, CURLINFO_HEADER_OUT, and CURLINFO_HEADER_IN messages.  Those are basically all of the text-based cases.  There are then the CURLINFO_DATA_* and CURLINFO_SSL_DATA_* binary messages, which at the moment don't seem as useful when simply dumped to stderr.  If we find a need in the future, we can certainly add our own custom debug function to route all of this information elsewhere, to handle those additional messages, etc.